packages/x-data-grid/src/components/GridSkeletonLoadingOverlay.tsx

```diff
@@ -0,0 +1,128 @@
+import * as React from 'react';
+import { styled } from '@mui/system';
+import useForkRef from '@mui/utils/useForkRef';
+import composeClasses from '@mui/utils/composeClasses';
+import { useGridApiContext } from '../hooks/utils/useGridApiContext';
+import { useGridRootProps } from '../hooks/utils/useGridRootProps';
+import {
+  gridColumnPositionsSelector,
+  gridColumnsTotalWidthSelector,
+  gridDimensionsSelector,
+  gridVisibleColumnDefinitionsSelector,
+  useGridApiEventHandler,
+  useGridSelector,
+} from '../hooks';
+import { GridEventListener } from '../models';
+import { DataGridProcessedProps } from '../models/props/DataGridProps';
+import { getDataGridUtilityClass } from '../constants/gridClasses';
+
+const colWidthVar = (index: number) => `--colWidth-${index}`;
+
+const SkeletonOverlay = styled('div', {
+  name: 'MuiDataGrid',
+  slot: 'SkeletonLoadingOverlay',
+  overridesResolver: (props, styles) => styles.skeletonLoadingOverlay,
+})({
+  display: 'grid',
+  width: 'max-content', // ensures overflow: hidden; does not cut off the x axis
+  minWidth: '100%', // ensures the filler column takes up the remaining space in a row
+  height: '100%',
+  overflow: 'hidden',
+  '& .MuiDataGrid-cellSkeleton': {
+    borderBottom: '1px solid var(--DataGrid-rowBorderColor)',
+  },
```

Comment:
We set the border on the top edge, this has to do with details panel placement, makes it easier. Rows also have a `--rowBorderColor` variable that can be different (for the first row), you should use that instead of `--DataGrid-rowBorderColor`.